juju/juju
# fix: migrate 3.6 to 4.0

We must ensure that 3.6 can migrate to 4.0 without issues once it lands into the 4.0 controller. This follows on from [2] which unlocks 3.6 to 4.0 from the 4.0 side. 3.6 is expected to be our LTS and 4.0 will be our new shiny. We must always ensure that we can migrate from there to 4.0.

Solving this is done to restore the errors from the APICall to ensure it's converted back to NotImplemented error (along with other errors). We don't do it for all results, but just the calls to the facade, as that's when it caught. For consistency, I've added to all uniter calls, with the aim that in new versions of Juju the API client calls should always do this.

The SLALevel and MeterStatus are both removed in 4.0[1] and so we need to handle when that occurs. The best way is to handle the error and then allow call sites to decide what to do.

Lastly, the meter-status and metric-sender workers will just complain and complain about the issue. So to prevent that, we trap the error in the manifold filter function and return nil. This will mark the worker as complete and won't come up again unless requested. We can't use `dependency.ErrUninstall` as other transitive dependencies bounce and chaos ensues.

The real solution to the problem is to then perform an `upgrade-model`, which will then move it to 4.0.x and everything will become normalized.

 1. https://github.com/juju/juju/pull/16942
 2.  https://github.com/juju/juju/pull/17596



## Checklist




## QA steps

For each path, you must also deploy into the model _after_ it's migrated, as it's at that point you'll see errors streaming out.

### 3.6 to 3.6

```sh
$ git checkout 3.6
$ make juju jujud jujud-controller
$ juju bootstrap lxd src36
$ juju add-model moveme
$ juju deploy ubuntu

$ juju bootstrap lxd dst36

$ juju switch src36
$ juju migrate src36:moveme dst36

$ juju deploy ubuntu ubuntux
$ juju debug-log -m dst36:moveme --replay
```

### Setup 3.6 to 4.0

```sh
$ git checkout 3.6
$ make juju jujud jujud-controller
$ juju bootstrap lxd src36
$ juju add-model moveme
$ juju deploy ubuntu

$ git checkout main
$ make juju jujud jujud-controller
$ juju bootstrap lxd dst40

$ juju switch src36
$ juju migrate src36:moveme dst40

$ juju deploy ubuntu ubuntux
$ juju debug-log -m dst40:moveme --replay
```

## Links


**Jira card:** [JUJU-6188](https://warthogs.atlassian.net/browse/JUJU-6188)



[JUJU-6188]: https://warthogs.atlassian.net/browse/JUJU-6188?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
